pyrealm/splash/splash.py

```diff
@@ -0,0 +1,278 @@
+
+        if wn_init is not None:
+            # Check the shape is the same as the shape of a slice along axis 0
+            if wn_init.shape != self.tc[0].shape:
+                raise ValueError("Incorrect shape in wn_init")
+            wn_start = wn_init
+        else:
+            wn_start = np.zeros_like(self.tc[0])
+
+        # Find a date one year into the future from the first calendar date.
+        # TODO - fix leap year handling and non Jan 1 starts
+
+        if self.tc.shape[0] < 365:
+            raise ValueError("Cannot equilibrate - less than one year of data")
+
+        # Run the equilibration loop
+        while (not equilibrated) and (n_iter < max_iter):
+            # Track the iterations
+            n_iter += 1
+
+            # Loop over the calendar object, updating the soil_moisture array
+            wn_day = wn_start
+            for day_idx in np.arange(366):
+                # Calculate aet, soil moisture and runoff:
+                _, wn_day, _ = self.estimate_daily_water_balance(
+                    previous_wn=wn_day, day_idx=day_idx
+                )
+
+            # Calculate the difference between the start of year soil moisture and the
+            # final day of the year and then update the start point to the end of year.
+            diff_sm = np.abs(wn_start - wn_day)
+            wn_start = wn_day
+
+            # Report if verbose
+            if verbose:
+                logger.info(f"Iteration: {n_iter}; maximum difference: {diff_sm.max()}")
+
+            if np.nanmax(diff_sm) <= max_diff:
+                equilibrated = True
+
+        # Check for convergence failure.
+        if not equilibrated:
+            raise RuntimeError(
+                f"Initial soil moisture did not converge within {n_iter} iterations:"
+                f"maximum absolute difference = {diff_sm.max()}"
+            )
+
+        return wn_start
+
+    def estimate_daily_water_balance(
+        self, previous_wn: NDArray, day_idx: Optional[int] = None
+    ) -> tuple[NDArray, NDArray, NDArray]:
+        r"""Estimate the daily water balance.
+
+        The daily soil moisture (wn) is estimated using the soil moisture from the
+        preceeding day, the precipitation for each day (``pn``) and the calculated
+        condensation (``cn``) and AET given the evaporative fluxes:
+
+        ..math::
+
+            \textrm{wn}_{t} = \textrm{wn}_{t-1} + \textrm{pn} +
+                              \textrm{cn} - \textrm{aet}
+
+        The calculated soil moisture is then partitioned into soil moisture and any
+        surplus runoff, given the maximum soil capacity for each observation. Negative
+        soil moisture values are replaced by zero.
+
+        By default, ``previous_wn`` is expected to provide estimates for all
+        observations across all days in the model, but ``day_idx`` can be set to provide
+        an estimate for only one particular day, for use in iterating over time series.
+
```

Comment:
Maybe add a sentence that the water in the bucket does not move vertically or horizontally (subsurface flow) to end up in the ocean.